dawn/src/dawn/IIR/ASTStmt.cpp

```diff
@@ -13,16 +13,89 @@
 //===------------------------------------------------------------------------------------------===//
 
 #include "dawn/IIR/ASTStmt.h"
+#include "dawn/IIR/Extents.h"
 #include <memory>
 
 namespace dawn {
 namespace iir {
 
-bool IIRStmtData::operator==(const IIRStmtData& rhs) { return StackTrace == rhs.StackTrace; }
+//===------------------------------------------------------------------------------------------===//
+//     IIRStmtData
+//===------------------------------------------------------------------------------------------===//
+
+IIRStmtData::IIRStmtData(const IIRStmtData& other) {
+  StackTrace = other.StackTrace ? std::make_optional(*other.StackTrace) : other.StackTrace;
+  CallerAccesses =
+      other.CallerAccesses ? std::make_optional(*other.CallerAccesses) : other.CallerAccesses;
+  CalleeAccesses =
+      other.CalleeAccesses ? std::make_optional(*other.CalleeAccesses) : other.CalleeAccesses;
+}
+
+bool IIRStmtData::operator==(const IIRStmtData& rhs) {
+  return StackTrace == rhs.StackTrace && CallerAccesses == rhs.CallerAccesses &&
+         CalleeAccesses == rhs.CalleeAccesses;
+}
 bool IIRStmtData::operator!=(const IIRStmtData& rhs) { return !(*this == rhs); }
 
 std::unique_ptr<ast::StmtData> IIRStmtData::clone() const {
   return std::make_unique<IIRStmtData>(*this);
 }
+
+//===------------------------------------------------------------------------------------------===//
+//     VarDeclStmtData
+//===------------------------------------------------------------------------------------------===//
+
+VarDeclStmtData::VarDeclStmtData(const VarDeclStmtData& other) : IIRStmtData(other) {
+  AccessID = other.AccessID ? std::make_optional(*other.AccessID) : other.AccessID;
+}
+
+bool VarDeclStmtData::operator==(const VarDeclStmtData& rhs) {
+  return IIRStmtData::operator==(rhs) && AccessID == rhs.AccessID;
+}
+bool VarDeclStmtData::operator!=(const VarDeclStmtData& rhs) { return !(*this == rhs); }
+
+std::unique_ptr<ast::StmtData> VarDeclStmtData::clone() const {
+  return std::make_unique<VarDeclStmtData>(*this);
+}
+
+//===------------------------------------------------------------------------------------------===//
+//     computeMaximumExtents
+//===------------------------------------------------------------------------------------------===//
+
+std::optional<Extents> computeMaximumExtents(Stmt& stmt, const int accessID) {
+  std::optional<Extents> extents;
+
+  const auto& callerAccesses = stmt.getData<IIRStmtData>().CallerAccesses;
+
+  if(callerAccesses->hasReadAccess(accessID) || callerAccesses->hasWriteAccess(accessID)) {
```

Comment:
80 might go into the then case, if callerAccess has no ReadAccess. In that case, extents stays empty and we go into then.

Maybe something like this would make it clearer?
```
bool writeAccess = callerAccesses->hasWriteAccess(accessID);
bool readAccess = callerAccesses->hasWriteAccess(accessID);
if (readAccess && writeAccess)
    extents = callerAccesses->getReadAccess(accessID)->merge(callerAccesses->getWriteAccess(accessID));
else if (readAccess)
    extents = callerAccesses->getReadAccess(accessID);
else if (writeAccess)
    extents = callerAccesses->getWriteAccess(accessID);
```

Reply:
Problem is that we need to work on a copy of extents as we cannot use non-cost `merge` on `const` Extents returned by `getReadAccess(accessID)` ... also `extents` is `std::optional`...
Can't come up with nothing better than:
```
if(callerAccesses->hasReadAccess(accessID))
    extents = std::make_optional(callerAccesses->getReadAccess(accessID));

  if(callerAccesses->hasWriteAccess(accessID)) {
    if(extents)
      extents->merge(callerAccesses->getWriteAccess(accessID));
    else
      extents = std::make_optional(callerAccesses->getWriteAccess(accessID));
  }
```